src/lib/storage/fsst_segment.cpp

```diff
@@ -0,0 +1,167 @@
+                            std::unique_ptr<const BaseCompressedVector>& compressed_offsets,
+                            pmr_vector<uint64_t>& reference_offsets, std::optional<pmr_vector<bool>>& null_values,
+                            uint64_t number_elements_per_reference_bucket, fsst_decoder_t& decoder)
+    : AbstractEncodedSegment{data_type_from_type<pmr_string>()},
+      _compressed_values{std::move(compressed_values)},
+      _compressed_offsets{std::move(compressed_offsets)},
+      _reference_offsets{std::move(reference_offsets)},
+      _null_values{std::move(null_values)},
+      _number_elements_per_reference_bucket{number_elements_per_reference_bucket},
+      _decoder{decoder},
+      _offset_decompressor{_compressed_offsets->create_base_decompressor()} {}
+
+template <typename T>
+AllTypeVariant FSSTSegment<T>::operator[](const ChunkOffset chunk_offset) const {
+  PerformanceWarning("operator[] used");
+  DebugAssert(chunk_offset < size(), "Passed chunk offset must be valid.");
+
+  const auto typed_value = get_typed_value(chunk_offset);
+  if (typed_value) {
+    return typed_value.value();
+  }
+  return NULL_VALUE;
+}
+
+template <typename T>
+uint64_t FSSTSegment<T>::get_offset(const ChunkOffset chunk_offset) const {
+  // For the number of values below 8 (number of reference offsets in encoder) we don't use reference offsets;
+  // In that case _number_elements_per_reference_bucket == 0.
+  // For the first bucket we don't use reference offsets either.
+  if (_number_elements_per_reference_bucket == 0 || chunk_offset < _number_elements_per_reference_bucket) {
+    return _offset_decompressor->get(chunk_offset);
+  }
+
+  // Calculate the corresponding reference offset index for the chunk_offset.
+  auto reference_offset_index = (chunk_offset / _number_elements_per_reference_bucket) - 1;
+  // Merge the last uncompleted bucket with the second last bucket.
+  reference_offset_index = std::min(reference_offset_index, _reference_offsets.size() - 1);
+
+  // Remove the "zig-zag" pattern and return the original offset.
+  return _offset_decompressor->get(chunk_offset) + _reference_offsets[reference_offset_index];
+}
+
+template <typename T>
+std::optional<T> FSSTSegment<T>::get_typed_value(const ChunkOffset chunk_offset) const {
+  if (_null_values) {
+    if (_null_values.value()[chunk_offset]) {
+      return std::nullopt;
+    }
+  }
+
+  // Calculate real offset with help of reference offset vector.
+  size_t real_offset = get_offset(chunk_offset);
+  size_t real_offset_next = get_offset(chunk_offset + 1);
+  size_t compressed_length = real_offset_next - real_offset;
+
+  // Note: we use const_cast in order to use fsst_decompress.
+  auto* compressed_pointer = const_cast<unsigned char*>(  // NOLINT(cppcoreguidelines-pro-type-const-cast)
+      _compressed_values.data() + real_offset);
+
+  // Since the max symbol length is 8, max uncompressed size is 8 * compressed_length.
+  size_t output_size = compressed_length * 8;
```

Comment:
Many `const`s missing.

Reply:
Name is `max_output_size` since it's not the actual size but rather an upper bound?